x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanStreamOutput.java

```diff
@@ -28,19 +41,34 @@
  */
 public final class PlanStreamOutput extends StreamOutput {
 
+    private final Map<Block, BytesReference> cachedBlocks = new IdentityHashMap<>();
```

Comment:
I have some concerns about using Block as the key since we'll need to read all values in hashCode and equals. However, if you're okay with that, I'm fine too.

Reply:
It's an `IdentityHashMap` - just the object itself.

Reply:
I'll leave a comment about how that's important. It's be expensive to use the `Block` as a regular key.